File: Data Structures & Algorithms/longest-common-subsequence/submission-0.cpp
class Solution {
public:
    int dp[1002][1002];
    int helper(string s1,string s2,int i,int j){

        if(i<0 || j< 0)
            return 0;
        
        if(dp[i][j] != -1)
            return dp[i][j];
        int res =0;
        
        if(s1[i]==s2[j])
            res = 1+ helper(s1,s2,i-1,j-1);
        else res = max(helper(s1,s2,i,j-1),helper(s1,s2,i-1,j));
        
        return dp[i][j]=res;    


    }
    int longestCommonSubsequence(string t1, string t2) {
        int m = t1.size();
        int n = t2.size();
        for(int i=0;i<t1.size();i++)
            for(int j=0;j<t2.size();j++)
                dp[i][j]=-1;

        return helper(t1,t2,m-1,n-1);        
    }
};

File: Data Structures & Algorithms/edit-distance/submission-0.java
class Solution {
    public int solve(int[][] dp,String word1,String word2,int i,int j){

        if(i<0 && j<0)
            return 0;
        else if(i<0 || j<0){
            return 1+ Math.max(i,j);
        }        

        if(dp[i][j]!=-1)
            return dp[i][j];

        int res;
        int case1,case2,case3;
        if(word1.charAt(i)==word2.charAt(j)){
            res = solve(dp,word1,word2,i-1,j-1);
        }else{

        // insert A character
            case1 = 1+solve(dp,word1,word2,i,j-1);
        // delete A character
            case2 = 1+solve(dp,word1,word2,i-1,j);
        // replace A character
            case3 = 1+solve(dp,word1,word2,i-1,j-1);

            res = Math.min(case1,Math.min(case2,case3));
        }

        return dp[i][j]=res;
               
        
    }
    public int minDistance(String word1, String word2) {
        int m = word1.length();
        int n = word2.length();

        int[][] dp = new int[m+1][n+1];

        for(int i=0;i<=m;i++)
            for(int j=0;j<=n;j++)
                dp[i][j]=-1;

        return solve(dp,word1,word2,m-1,n-1); 

    }
}
